go/oasis-node/cmd/debug/txsource/workload/governance.go

```diff
@@ -0,0 +1,479 @@
+	gracefulExit context.Context,
+	rng *rand.Rand,
+	conn *grpc.ClientConn,
+	cnsc consensus.ClientBackend,
+	sm consensus.SubmissionManager,
+	fundingAccount signature.Signer,
+	validatorEntities []signature.Signer,
+) error {
+	var err error
+
+	// Initialize state.
+	g.BaseWorkload.Init(cnsc, sm, fundingAccount)
+	g.rng = rng
+	g.ctx = context.Background()
+
+	g.parameters, err = cnsc.Governance().ConsensusParameters(g.ctx, consensus.HeightLatest)
+	if err != nil {
+		return fmt.Errorf("error querying governance consensus parameters: %w", err)
+	}
+
+	g.governance = cnsc.Governance()
+
+	g.validatorEntities = validatorEntities
+	if len(g.validatorEntities) == 0 {
+		return fmt.Errorf("workload requires validator entities")
+	}
+
+	g.proposerAccounts = make([]*struct {
+		signer  signature.Signer
+		address staking.Address
+		nonce   uint64
+	}, numProposerAccounts)
+	fac := memorySigner.NewFactory()
+	for i := range g.proposerAccounts {
+		var signer signature.Signer
+		signer, err = fac.Generate(signature.SignerEntity, rng)
+		if err != nil {
+			return fmt.Errorf("memory signer factory Generate account %d: %w", i, err)
+		}
+		g.proposerAccounts[i] = &struct {
+			signer  signature.Signer
+			address staking.Address
+			nonce   uint64
+		}{
+			signer:  signer,
+			address: staking.NewAddress(signer.Public()),
+		}
+		if err = g.TransferFunds(
+			g.ctx,
+			fundingAccount,
+			g.proposerAccounts[i].address,
+			g.parameters.MinProposalDeposit.ToBigInt().Uint64()*1000,
+		); err != nil {
+			return fmt.Errorf("account funding failure: %w", err)
+		}
+	}
+
+	g.ensureYesVote = make(map[uint64]bool)
+
+	// Main workload loop.
+	for {
+		select {
+		case <-time.After(1 * time.Second):
+		case <-gracefulExit.Done():
+			g.Logger.Debug("time's up")
+			return nil
+		}
+
+		var epoch epochtime.EpochTime
+		epoch, err = cnsc.GetEpoch(g.ctx, consensus.HeightLatest)
+		if err != nil {
+			return fmt.Errorf("querying epoch: %w", err)
+		}
+
+		// Ensure epoch transition won't happen for the next 10 blocks. Otherwise
+		// if the epoch transition happens during the loop iteration, we can have
+		// outdated local state.
+		var epoch2 epochtime.EpochTime
+		var block *consensus.Block
+		block, err = cnsc.GetBlock(g.ctx, consensus.HeightLatest)
+		if err != nil {
+			return fmt.Errorf("querying block: %w", err)
+		}
+		epoch2, err = cnsc.GetEpoch(g.ctx, block.Height+10)
+		if err != nil {
+			return fmt.Errorf("querying epoch: %w", err)
+		}
+		if epoch2 != epoch {
+			// Epoch transition soon.
+			continue
+		}
+
+		// Epoch transition - update the local state accordingly.
+		if epoch > g.currentEpoch {
+			if err = g.handleEpochTransition(epoch); err != nil {
+				return fmt.Errorf("handle epoch transition: %w", err)
+			}
+
+			// Make sure no pending upgrade will go through.
+			// XXX: this makes sure that any pending upgrades that are about to be executed are
+			// canceled. When txsource suite supports handling upgrades mid-run, remove this part.
```

Comment:
Is there an issue for tracking this?

Reply:
will open one

Reply:
opened https://github.com/oasisprotocol/oasis-core/issues/3609